src/cryptonote_core/blockchain.cpp

```diff
@@ -1102,7 +1136,11 @@ size_t Blockchain::recalculate_difficulties(boost::optional<uint64_t> start_heig
 //------------------------------------------------------------------
 std::vector<time_t> Blockchain::get_last_block_timestamps(unsigned int blocks) const
 {
-  CRITICAL_REGION_LOCAL(m_blockchain_lock);
+  m_blockchain_transaction.start_write();
```

Comment:
Why are all these `start_write()` even for functions which only read data?

Reply:
Consider this as prototype/experimental PR, which does have bugs.

This is a symptom of a bigger problem we have. Imagine this scenario: Functions calls: `A`->`B`->`C`->`D`.

`A` is a write transaction. But `B`, `C` and `D` are read transactions. We have to acquire a write lock for `B`, `C`, and `D` since read and write transactions in this implementation are exclusive. (I am fixing it).